src/engine_base.h

```diff
@@ -142,15 +142,21 @@ struct Engine : EnginePool::PoolItem<&_engine_pool> {
 
 	uint32 GetGRFID() const;
 
+	struct EngineTypeFilter {
+		VehicleType vt;
+
+		bool operator() (size_t index) { return Engine::Get(index)->type == this->vt; }
+	};
+
 	/**
 	 * Returns an iterable ensemble of all valid engines of the given type
 	 * @param vt the VehicleType for engines to be valid
 	 * @param from index of the first engine to consider
 	 * @return an iterable ensemble of all valid engines of the given type
 	 */
-	static Pool::IterateWrapper<Engine> IterateType(VehicleType vt, size_t from = 0)
+	static Pool::IterateWrapperFiltered<Engine, EngineTypeFilter> IterateType(VehicleType vt, size_t from = 0)
 	{
-		return Pool::IterateWrapper<Engine>(from, [vt](size_t index) { return Engine::Get(index)->type == vt; });
+		return Pool::IterateWrapperFiltered<Engine, EngineTypeFilter>(from, EngineTypeFilter{ vt });
```

Comment:
Without having actually tried it, but a lambda is supposed to implicitly convert to a functor.

Reply:
Unfortunately you cannot use a lambda as a template parameter, and the full typename needs to be in the return type
Very broadly, lambdas only usefully convert to functors when using them as a parameter to a templated function call.